### ISSUE TITLE
option_if_let_else suggestion inapplicable when mutable borrow or owned value is used in both branches

### ISSUE THREAD

**Issue by Arnavion** [rust-lang/rust-clippy#5822]
```rust
#[deny(clippy::option_if_let_else)]
pub fn foo(o: Option<i32>, mut f: impl FnMut(i32)) {
    if let Some(i) = o {
        f(i)
    }
    else {
        f(5)
    }
}
```

Apart from having the same problem as #5821, in this particular case even the suggestion of `o.map_or_else(|| f(5), |i| f(i))` would be wrong because `f` is being used as a `&mut`, so it cannot be used with two closures at the same time. This also happens for `f: impl FnOnce(i32)`.

I don't think there's any better way to write this code using any of the `Option` combinators, so it seems to be the lint should not fire at all.

### Meta

- `cargo clippy -V`: `clippy 0.0.212 (39d5a61 2020-07-17)`
- `rustc -Vv`:
  ```
  rustc 1.47.0-nightly (39d5a61f2 2020-07-17)
  binary: rustc
  commit-hash: 39d5a61f2e4e237123837f5162cc275c2fd7e625
  commit-date: 2020-07-17
  host: x86_64-unknown-linux-gnu
  release: 1.47.0-nightly
  LLVM version: 10.0
  ```

**Comment by rail-rain** [comment]
I've found a real world example inside Clippy where presumably this bug led to an incorrect use of `crate::util::snippet_with_applicability`, which slipped into the code base [along with the lint itself](https://github.com/rust-lang/rust-clippy/pull/5301/commits/5150277a4f765bb113e163adc7eb495dcbb57129). I'm not sure whether we should immediately revert the change with `#[allow(clippy::option_if_let_else)]` for the time being or can leave it until this issue gets fixed though.

https://github.com/rust-lang/rust-clippy/blob/23db5426e6c90a18700e6bf30af537e5bf6b14c7/clippy_lints/src/write.rs#L299-L306

**Comment by ebroto** [comment]
Nice catch! I would say let's allow the lint for now and fix that.

BTW wouldn't this be a nice candidate for a lint? I can't think now of a case where you actually want to pass a `&mut` to a temporary to a function.

**Comment by rail-rain** [comment]
OK, I made a PR for it (and it's just quickly merged while I'm writing this). Thanks for the reply!

Regarding an new lint, it makes sense for me; considering a function taking a mutable reference probably means the reference is some kind of an "out" parameter.

**Comment by brightly-salty** [comment]
I still have this issue on nightly, I don't know if it's been fixed or not.

**Comment by edward-shen** [comment]
I can confirm this issue still exists as I'm running into this on cargo 1.54.0-nightly (e51522ab3 2021-05-07).
